File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/handler/check/impl/UnionExtractValidator.java
package ru.ioleksiv.telegram.bot.core.controller.handler.check.impl;

import ru.ioleksiv.telegram.bot.core.controller.handler.check.Validator;
import ru.ioleksiv.telegram.bot.core.controller.handler.unpack.Unpacker;

import java.util.ArrayList;
import java.util.Collection;

public class UnionExtractValidator<T> implements Validator<T> {

    private final Collection<Validator<T>> validatorCollection = new ArrayList<>();

    public <OUT> void add(Unpacker<T, OUT> unpacker, Validator<OUT> validator) {
        validatorCollection.add(new ExtractValidator<>(unpacker, validator));
    }

    @Override
    public boolean check(T argument) {
        return validatorCollection.stream().allMatch(checker -> checker.check(argument));
    }
}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/handler/check/impl/UnionExtractValidator.kt
package io.github.oleksivio.telegram.bot.core.controller.handler.check.impl

import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator
import io.github.oleksivio.telegram.bot.core.controller.handler.unpack.UnpackerFunction
import java.util.*

class UnionExtractValidator<T> : Validator<T> {
    override fun invoke(checkedValue: T): Boolean {
        return validatorCollection.stream().allMatch { checker -> checker.invoke(checkedValue) }
    }

    private val validatorCollection = ArrayList<Validator<T>>()

    fun <OUT> add(unpacker: UnpackerFunction<T, OUT>, validator: Validator<OUT>) {
        validatorCollection.add(ExtractValidator(unpacker, validator))
    }


}
